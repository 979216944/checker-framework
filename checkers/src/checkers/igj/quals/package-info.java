/**
 * Contains qualifiers for type annotations used by the IGJ checker.
 *
 * @see checkers.igj.IGJChecker
 *
 * @manual #igj-checker IGJ Checker
 */
package checkers.igj.quals;
