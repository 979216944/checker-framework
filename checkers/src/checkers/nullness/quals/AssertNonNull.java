package checkers.nullness.quals;

import java.lang.annotation.*;

import checkers.nullness.NullnessChecker;

/**
 * Indicates that the method throws a {@link Throwable} if any of its
 * parameters is a {@code null}.
 *
 * @see NonNull
 * @see NullnessChecker
 * @manual #nullness-checker Nullness Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface AssertNonNull {
}
