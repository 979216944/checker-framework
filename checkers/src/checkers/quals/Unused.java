package checkers.quals;

import static java.lang.annotation.ElementType.*;

import java.lang.annotation.*;

/**
 * Declares that the field may not be accessed if the receiver is of the
 * specified qualifier type.
 *
 * The checker that type-checks the {@code when} element value qualifier
 * verifies that property.
 *
 * <p><b>Example</b>
 * Consider a class, {@code Table}, with a locking field, {@code lock}.  The
 * lock is used when a {@code Table} instance is shared across threads.  When
 * running in a local thread, the {@code lock} field ought not to be used.
 *
 * You can declare this behavior in the following way:
 *
 * <pre><code>
 * class Table {
 *   private @Unused(when=LocalToThread.class) final Lock lock;
 *   ...
 * }
 * </code></pre>
 *
 * The appropriate checker (which type-checks the specified qualifier) is to
 * issue an error whenever the lock is used in a {@code LocalToThread}, like
 * in the following case:
 *
 * <pre><code>
 *   final @LocalToThread Table table = ....;
 *   table.lock.lock();
 * </code></pre>
 *
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Unused {
    Class<? extends Annotation> when();
}
