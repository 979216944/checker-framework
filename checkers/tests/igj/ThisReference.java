import checkers.igj.quals.*;

@I
public class ThisReference {

    // Receiving methods
    public static void isRO(@ReadOnly ThisReference obj) { }
    public static void isMutable(@Mutable ThisReference obj) { }
    public static void isImmutable(@Immutable ThisReference obj) { }

    public void testRO() @ReadOnly {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this); // should emit error
    }

    public void testAssignsFields () @AssignsFields {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this); // should emit error
    }

    public void testMutable() @Mutable {
        isRO(this);
        isMutable(this);
        isImmutable(this); // should emit error
    }

    public void testImmutable() @Immutable {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this);
    }
}


@Immutable
class ImmutableThisTest extends ThisReference {
    public void testRO() @ReadOnly {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this);
    }

    public void testAssignsFields () @AssignsFields {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this);
    }

    public void testImmutable() @Immutable {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this);
    }
}

@Mutable
class MutableThisTest extends ThisReference {
    public void testRO() @ReadOnly {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this); // should emit error
    }

    public void testAssignsFields () @AssignsFields {
        isRO(this);
        isMutable(this);   // should emit error
        isImmutable(this); // should emit error
    }

    public void testMutable() @Mutable {
        isRO(this);
        isMutable(this);
        isImmutable(this); // should emit error
    }
}