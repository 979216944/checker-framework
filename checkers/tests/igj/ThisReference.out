ThisReference.java:13: (type.incompatible)
ThisReference.java:14: (type.incompatible)
ThisReference.java:19: (type.incompatible)
ThisReference.java:20: (type.incompatible)
ThisReference.java:26: (type.incompatible)
ThisReference.java:31: (type.incompatible)
ThisReference.java:41: (type.incompatible)
ThisReference.java:47: (type.incompatible)
ThisReference.java:53: (type.incompatible)
ThisReference.java:62: (type.incompatible)
ThisReference.java:63: (type.incompatible)
ThisReference.java:68: (type.incompatible)
ThisReference.java:69: (type.incompatible)
ThisReference.java:75: (type.incompatible)
